internal/ssm/registration.go

```diff
@@ -0,0 +1,143 @@
+	DeregisterManagedInstance(ctx context.Context, params *awsSsm.DeregisterManagedInstanceInput, optFns ...func(*awsSsm.Options)) (*awsSsm.DeregisterManagedInstanceOutput, error)
+}
+
+func WithInstallRoot(installRoot string) SSMRegistrationOption {
+	return func(c *SSMRegistration) {
+		c.installRoot = installRoot
+	}
+}
+
+func Deregister(ctx context.Context, registration ISSMRegistration, logger *zap.Logger) error {
+	instanceId, err := registration.GetManagedHybridInstanceId()
+
+	// If uninstall is being run just after running install and before running init
+	// SSM would not be fully installed and registered, hence it's not required to run
+	// deregister instance.
+	if err != nil && os.IsNotExist(err) {
+		logger.Info("Skipping SSM deregistration - node is not registered")
+		return nil
+	}
+	if err != nil {
+		return errors.Wrapf(err, "reading ssm registration file")
+	}
+
+	ssmClient, err := registration.SSMClient(ctx)
+	if err != nil {
+		return errors.Wrapf(err, "getting ssm client")
+	}
+	managed, err := isInstanceManaged(ssmClient, instanceId)
+	if err != nil {
+		return errors.Wrapf(err, "getting managed instance information")
+	}
+
+	// Only deregister the instance if init/ssm init was run and
+	// if instances is actively listed as managed
+	if managed {
+		if err := deregister(ssmClient, instanceId); err != nil {
+			return errors.Wrapf(err, "deregistering ssm managed instance")
+		}
+	}
+	return nil
+}
+
+func (r *SSMRegistration) getManagedHybridInstanceIdAndRegion() (string, string, error) {
+	data, err := os.ReadFile(r.RegistrationFilePath())
+	if err != nil {
+		return "", "", err
+	}
+
+	var registration HybridInstanceRegistration
+	err = json.Unmarshal(data, &registration)
+	if err != nil {
+		return "", "", err
+	}
+	return registration.ManagedInstanceID, registration.Region, nil
+}
+
+func (r *SSMRegistration) GetManagedHybridInstanceId() (string, error) {
+	data, err := os.ReadFile(r.RegistrationFilePath())
+	if err != nil {
+		return "", err
+	}
+
+	var registration HybridInstanceRegistration
+	err = json.Unmarshal(data, &registration)
+	if err != nil {
+		return "", err
+	}
+	return registration.ManagedInstanceID, nil
+}
+
+func (r *SSMRegistration) isRegistered() (bool, error) {
+	_, err := r.GetManagedHybridInstanceId()
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("reading ssm registration file: %w", err)
+	}
+	return true, nil
+}
+
+// RegistrationFilePath returns the path to the SSM registration file
+// If installRoot is not set, it will return the path starting from the disk root
+func (r *SSMRegistration) RegistrationFilePath() string {
+	return filepath.Join(r.installRoot, registrationFilePath)
+}
+
+func (r *SSMRegistration) SSMClient(ctx context.Context) (SSMClient, error) {
```

Comment:
I would just offer a method that returns the region
That can use to build the client or for other things but it decouples reading the region from a registration from building a client for a particular region